src/transaction/TransactionController.ts

```diff
@@ -1273,6 +1241,97 @@ export class TransactionController extends BaseController<
     );
   }
 
+  /**
+   * Method to verify the state of a transaction using the Blockchain as a source of truth
+   * @param meta Local transaction to verify data in blockchain
+   * @returns Promise with [TransactionMeta, boolean]
+   */
+  private async blockchainTransactionStateReconciler(
+    meta: TransactionMeta,
+  ): Promise<[TransactionMeta, boolean]> {
+    const { status, transactionHash } = meta;
+    switch (status) {
+      case TransactionStatus.confirmed:
+        const txReceipt = await query(this.ethQuery, 'getTransactionReceipt', [
+          transactionHash,
+        ]);
+
+        if (!txReceipt) {
+          return [meta, false];
+        }
+
+        meta.verifiedOnBlockchain = true;
+        meta.transaction.gasUsed = txReceipt.gasUsed;
+
+        // According to the Web3 docs:
+        // TRUE if the transaction was successful, FALSE if the EVM reverted the transaction.
+        if (Number(txReceipt.status) === 0) {
+          const error: Error = new Error(
+            'Transaction failed. The transaction was reversed',
+          );
+          this.failTransaction(meta, error);
+          return [meta, false];
+        }
+
+        return [meta, true];
+      case TransactionStatus.submitted:
+        const txObj = await query(this.ethQuery, 'getTransactionByHash', [
+          transactionHash,
+        ]);
+
+        if (!txObj) {
+          const receiptShowsFailedStatus = await this.checkTxReceiptStatusIsFailed(
```

Comment:
In what circumstance would  `getTransactionReceipt` succeed where `getTransactionByHash` failed?

Reply:
It would be a pretty rare situation where one succeed and the other one failed, but this logic is a kind of safe net in case `getTransactionByHash` fails. If `getTransactionReceipt` also fails it would mean that there was a connectivity issue or something similar and we shouldn't trust that response, using only the `getTransactionByHash` response the controller would update the transaction status based on a query that failed because of a X reason and the app would display misleading data to the user.

**More information on this case**
This bug happened to me under some conditions, like poor connectivity, I wasn't able to determined all the circumstances and the mobile QA engineer (@cortisiko) wasn't able to reproduce the bug.

The situation that was occurring was that in some moment I would get a null object as response from `getTransactionByHash` and the tx would be updated as failed and it would skip the polling loop from this point on, but in Etherscan it would show up as confirmed. After I added the second check with `getTransactionReceipt` I haven't been able to reproduce the bug.

I'm taking a deeper look into this case and gathering more information to conclude if this solution should be updated. Also, I want to point out that I'm taking a look `eth-block-tracker` to use it a future refactor to the transaction controller state management logic.

Reply:
As always, I'm open to any suggestions or recommendations 👍

Reply:
Even a connectivity issue shouldn't make `getTransactionByHash` return `null`. I'd expect it to throw in that circumstance. It should only return `null` if Infura successfully received the response and replied that they don't recognize the hash.

If you're sure that `getTransactionByHash` returned `null` for a transaction that later showed up as confirmed, it might have been a fluke of timing. Infura might have legitimately dropped the transaction, only for another node to mine a block that included it. If that's the case, then the bug you saw remains, just it's less likely to encounter it because we're making two calls rather than one. I would bet that calling `getTransactionByHash` a second time instead of `getTransactionReceipt` would have identical results.

We had reports of this on the extension as well, and we addressed it by leaving the transaction as "pending" until we saw it was dropped for 3 consecutive blocks. Only after 3 blocks of Infura reporting it as dropped would we tell the user it was dropped. This seemed to reduce the chances of this problem occurring, though I think it's still possible (albeit very rare).

Reply:
A more robust solution for this probably deserves its own PR, so I certainly don't want to block on that here. But we ought to at least include some comments about why this is done in this manner, because it's really not obvious from reading the code.

Reply:
Thanks for the insights @Gudahtt! I have been taking a deeper look at this to get to the root cause and @ibrahimtaveras00 also saw this behaviour. I will add the comments and start a new development to update this solution.